examples/python/learning_pytorch.py

```diff
@@ -48,230 +47,250 @@
+            nn.ReLU()
+        )
+        self.conv2 = nn.Sequential(
+            nn.Conv2d(8, 8, kernel_size=3, stride=2, bias=False),
+            nn.BatchNorm2d(8),
+            nn.ReLU()
+        )
+
+        self.state_fc = nn.Sequential(
+            nn.Linear(96, 64),
+            nn.ReLU(),
+            nn.Linear(64, 1)
+        )
+
+        self.advantage_fc = nn.Sequential(
+            nn.Linear(96, 64),
+            nn.ReLU(),
+            nn.Linear(64, available_actions_count)
+        )
 
     def forward(self, x):
-        x = F.relu(self.conv1(x))
-        x = F.relu(self.conv2(x))
+        x = torch.from_numpy(x).float().to(DEVICE)
+        x = self.conv1(x)
+        x = self.conv2(x)
         x = x.view(-1, 192)
-        x = F.relu(self.fc1(x))
-        return self.fc2(x)
-
-criterion = nn.MSELoss()
-
-
-def learn(s1, target_q):
-    s1 = torch.from_numpy(s1)
-    target_q = torch.from_numpy(target_q)
-    s1, target_q = Variable(s1), Variable(target_q)
-    output = model(s1)
-    loss = criterion(output, target_q)
-    # compute gradient and do SGD step
-    optimizer.zero_grad()
-    loss.backward()
-    optimizer.step()
-    return loss
-
-def get_q_values(state):
-    state = torch.from_numpy(state)
-    state = Variable(state)
-    return model(state)
-
-def get_best_action(state):
-    q = get_q_values(state)
-    m, index = torch.max(q, 1)
-    action = index.data.numpy()[0]
-    return action
-
-
-def learn_from_memory():
-    """ Learns from a single transition (making use of replay memory).
-    s2 is ignored if s2_isterminal """
-
-    # Get a random minibatch from the replay memory and learns from it.
-    if memory.size > batch_size:
-        s1, a, s2, isterminal, r = memory.get_sample(batch_size)
-
-        q = get_q_values(s2).data.numpy()
-        q2 = np.max(q, axis=1)
-        target_q = get_q_values(s1).data.numpy()
-        # target differs from q only for the selected action. The following means:
-        # target_Q(s,a) = r + gamma * max Q(s2,_) if isterminal else r
-        target_q[np.arange(target_q.shape[0]), a] = r + discount_factor * (1 - isterminal) * q2
-        learn(s1, target_q)
-
-
-def perform_learning_step(epoch):
-    """ Makes an action according to eps-greedy policy, observes the result
-    (next state, reward) and learns from the transition"""
-
-    def exploration_rate(epoch):
-        """# Define exploration rate change over time"""
-        start_eps = 1.0
-        end_eps = 0.1
-        const_eps_epochs = 0.1 * epochs  # 10% of learning time
-        eps_decay_epochs = 0.6 * epochs  # 60% of learning time
-
-        if epoch < const_eps_epochs:
-            return start_eps
-        elif epoch < eps_decay_epochs:
-            # Linear decay
-            return start_eps - (epoch - const_eps_epochs) / \
-                               (eps_decay_epochs - const_eps_epochs) * (start_eps - end_eps)
+        x1 = x[:, :96]  # input for the net to calculate the state value
```

Comment:
This is sensitive to any image-size modifications. I know this breaks the "dueling DQN" part a little, but you could feed the same `x` to both `state_fc` and `advantage_fc` directly, and it should still work.

Reply:
So is the line above (x = x.view(-1, 192))  and the original code though. As long as I use any kind of flatten instead of global pooling this will always depend on the image size. I could remove this dependency by using x.shape[0] // 2 but the dependency on image size is still there there because one needs to specify it in the flatten operation and the number of input neurons.

Reply:
Ah right, good point on the static FC size! I am too used to automatically inferring it I forgot it does not happen automagically in PyTorch. We can leave it as it is for the sake of the example :)